### PR TITLE
Continue adding block-spares support

### PR DESCRIPTION
**Pull request by mganahl**

### CODE REVIEW

File: tensornetwork/block_sparse/block_tensor.py [hunk 1]
```diff
@@ -234,3 +233,215 @@ def compute_num_nonzero(charges: List[BaseCharge], flows: List[bool]) -> int:
...
+        [(row_ind.charge_labels == n) for n in range(num_blocks)]).reshape(
+            num_blocks, row_ind.dim)
+  else:
+    # faster method for large number of blocks
+    row_locs = np.zeros([num_blocks, row_ind.dim], dtype=bool)
+    row_locs[row_ind
+             .charge_labels, np.arange(row_ind.dim)] = np.ones(
+                 row_ind.dim, dtype=bool)
+
+  # block_dims = np.array([row_degen[row_to_block],col_degen[col_to_block]], dtype=np.uint32)
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/block_tensor.py]: Delete commented out code.

File: tensornetwork/block_sparse/block_tensor.py [hunk 2]
```diff
@@ -234,3 +233,215 @@ def compute_num_nonzero(charges: List[BaseCharge], flows: List[bool]) -> int:
...
+    for n in range(left_ind.num_unique):
+      temp_label = new_comb_labels[n, right_ind.charge_labels]
+      temp_keep = temp_label >= 0
+      reduced_rows[n] = temp_label[temp_keep]
+      row_locs[n] = np.where(temp_keep)[0]
+
+    reduced_labels = np.concatenate(
+        [reduced_rows[n] for n in left_ind.charge_labels])
+    reduced_locs = np.concatenate([
+        n * right_ind.dim + row_locs[left_ind.charge_labels[n]]
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/block_tensor.py]: Looks like most of this code is duplicated. Is there a reason for that?

File: tensornetwork/block_sparse/charge.py [hunk 3]
```diff
@@ -21,6 +21,8 @@
 from typing import List, Optional, Type, Any, Union
 
 
+#TODO: change from column to row order for unique labels
+#TODO: implement more efficient unique function
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/charge.py]: Label TODOs with your username

Example:
`TODO(chaseriley):`

File: tensornetwork/block_sparse/charge.py [hunk 4]
```diff
@@ -144,13 +146,15 @@ def __len__(self):
 
   def __eq__(self,
              target_charges: Union[np.ndarray, "BaseCharge"]) -> np.ndarray:
-
+    #FIXME: calling np.unique can cause significan overhead in some cases
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/charge.py]: `FIXME(<YOUR_USERNAME>)`

**Comment by Thenerdstation** [comment]
lmk when you're ready.

**Comment by mganahl** [comment]
should be good to go!

**Comment by mganahl** [comment]
I have the next one ready as well

File: tensornetwork/block_sparse/block_tensor.py [hunk 5]
```diff
@@ -234,3 +233,254 @@ def compute_num_nonzero(charges: List[BaseCharge], flows: List[bool]) -> int:
...
+
+  #     reduced_labels = np.concatenate(
+  #         [reduced_rows[n] for n in left_ind.charge_labels])
+  #     reduced_locs = np.concatenate([
+  #         row_pos[n] + row_locs[left_ind.charge_labels[n]]
+  #         for n in range(left_ind.dim)
+  #     ])
+  #   else:
+  #     # reduce combined qnums to include only those in target_charges
+  #     # reduced_rows = [0] * left_ind.num_unique
```
Comment (Thenerdstation) [inline tensornetwork/block_sparse/block_tensor.py]: Delete commented out code.

Comment (mganahl) [reply tensornetwork/block_sparse/block_tensor.py]: d'oh, removed it

**Comment by mganahl** [comment]
I think probably ~15 or so. I'm free to work fulltime on this again, I think I can have most of it prepared during this week

**Pull request merged**